File: apps/mobile/src/lib/api/detectBoard.ts
import { base64ToDataUrl, apiFetch } from "@/lib/api/client";
import { WARP_PREPROCESS } from "@/lib/config";
import { BoardMatrixCell, DetectionDebug, DetectionResult, MlSquareDebug, SquareInfo } from "@/types";

interface DetectBoardApiResponse {
  job_id: string;
  corners: { x: number; y: number }[];
  confidence: number;
  original_width: number;
  original_height: number;
  output_width: number;
  output_height: number;
  warped_image_base64: string;
  fen: string | null;
  interactive_fen: string | null;
  fen_confidence: number | null;
  fen_valid: boolean | null;
  board_ready: boolean;
  board_matrix: { label: string; confidence: number }[][] | null;
  orientation: string | null;
  debug_overlay_base64: string | null;
  debug_montage_base64: string | null;
  debug: {
    original_base64?: string | null;
    detected_lines_base64?: string | null;
    intersections_base64?: string | null;
    mesh_base64?: string | null;
    rectified_board_base64?: string | null;
    rectified_upscaled_base64?: string | null;
    square_extraction_base64?: string | null;
    crop_quality_base64?: string | null;
    occupancy_base64?: string | null;
    occupancy_detail_base64?: string | null;
    classifier_confidence_base64?: string | null;
    ml_occupancy_base64?: string | null;
    ml_piece_top1_base64?: string | null;
    ml_onnx_occ_crops_base64?: string | null;
    ml_onnx_piece_crops_base64?: string | null;
    ml_detail_base64?: string | null;
    fen_candidates_base64?: string | null;
    mesh_quality_base64?: string | null;
    final_board_base64?: string | null;
    grid_debug_extreme_base64?: string | null;
    dataset_squares_base64?: string | null;
    rectified_grid_base64?: string | null;
    corners_original_base64?: string | null;
    corners_warped_base64?: string | null;
    rectified_preprocessed_base64?: string | null;
  } | null;
  squares: {
    name: string;
    filename: string;
    cell_bbox: number[];
    crop_bbox: number[];
    url: string;
    label?: string;
    confidence?: number;
    occupied?: boolean;
  }[];
  processing_ms: number;
  metadata: Record<string, unknown>;
}

function mapDebug(raw: DetectBoardApiResponse["debug"]): DetectionDebug | null {
  if (!raw) return null;
  const steps: DetectionDebug = {};
  const map: [keyof DetectionDebug, keyof NonNullable<typeof raw>][] = [
    ["original", "original_base64"],
    ["detectedLines", "detected_lines_base64"],
    ["intersections", "intersections_base64"],
    ["mesh", "mesh_base64"],
    ["rectifiedBoard", "rectified_board_base64"],
    ["rectifiedUpscaled", "rectified_upscaled_base64"],
    ["squareExtraction", "square_extraction_base64"],
    ["cropQuality", "crop_quality_base64"],
    ["occupancy", "occupancy_base64"],
    ["occupancyDetail", "occupancy_detail_base64"],
    ["classifierConfidence", "classifier_confidence_base64"],
    ["mlOccupancy", "ml_occupancy_base64"],
    ["mlPieceTop1", "ml_piece_top1_base64"],
    ["mlOnnxOccCrops", "ml_onnx_occ_crops_base64"],
    ["mlOnnxPieceCrops", "ml_onnx_piece_crops_base64"],
    ["mlDetail", "ml_detail_base64"],
    ["fenCandidates", "fen_candidates_base64"],
    ["meshQuality", "mesh_quality_base64"],
    ["finalBoard", "final_board_base64"],
    ["gridDebugExtreme", "grid_debug_extreme_base64"],
    ["rectifiedGrid", "rectified_grid_base64"],
    ["cornersOriginal", "corners_original_base64"],
    ["cornersWarped", "corners_warped_base64"],
    ["rectifiedPreprocessed", "rectified_preprocessed_base64"],
  ];
  for (const [target, source] of map) {
    const b64 = raw[source];
    if (b64) steps[target] = base64ToDataUrl(b64);
  }
  return Object.keys(steps).length > 0 ? steps : null;
}

function mapSquares(raw: DetectBoardApiResponse["squares"]): SquareInfo[] {
  return raw.map((sq) => ({
    name: sq.name,
    filename: sq.filename,
    cellBbox: sq.cell_bbox as SquareInfo["cellBbox"],
    cropBbox: sq.crop_bbox as SquareInfo["cropBbox"],
    label: sq.label,
    confidence: sq.confidence,
    occupied: sq.occupied,
  }));
}

function mapBoardMatrix(
  raw: DetectBoardApiResponse["board_matrix"],
): BoardMatrixCell[][] | undefined {
  if (!raw) return undefined;
  return raw.map((row) =>
    row.map((cell) => ({ label: cell.label, confidence: cell.confidence })),
  );
}

export async function detectBoard(
  file: File,
  originalUrl: string,
  endpoint = "/detect-board",
): Promise<DetectionResult> {
  const form = new FormData();
  form.append("file", file);

  const params = new URLSearchParams();
  if (endpoint.includes("lc2fen") && WARP_PREPROCESS && WARP_PREPROCESS !== "none") {
    params.set("warp_preprocess", WARP_PREPROCESS);
  }
  const qs = params.toString();
  const url = qs ? `${endpoint}?${qs}` : endpoint;

  const data = await apiFetch<DetectBoardApiResponse>(url, {
    method: "POST",
    body: form,
  });

  if (!data.fen && !data.squares?.length) {
    throw new Error("Classification pipeline did not return a FEN or square labels");
  }

  return {
    jobId: data.job_id,
    confidence: data.confidence,
    originalUrl,
    warpedUrl: base64ToDataUrl(data.warped_image_base64),
    overlayUrl: data.debug_overlay_base64
      ? base64ToDataUrl(data.debug_overlay_base64)
      : "/mock/overlay-board.svg",
    montageUrl: data.debug_montage_base64
      ? base64ToDataUrl(data.debug_montage_base64)
      : undefined,
    corners: data.corners,
    fen: data.fen ?? "",
    interactiveFen: data.interactive_fen,
    fenConfidence: data.fen_confidence ?? undefined,
    fenValid: data.fen_valid ?? undefined,
    boardReady: data.board_ready,
    boardMatrix: mapBoardMatrix(data.board_matrix),
    orientation: data.orientation ?? undefined,
    originalWidth: data.original_width,
    originalHeight: data.original_height,
    outputWidth: data.output_width,
    outputHeight: data.output_height,
    squares: mapSquares(data.squares),
    processingMs: data.processing_ms,
    debug: mapDebug(data.debug),
    mlDebug: mapMlDebug(data.metadata),
    metadata: data.metadata,
  };
}

function mapMlDebug(metadata: Record<string, unknown>): Record<string, MlSquareDebug> | null {
  const raw = metadata.ml_debug as { squares?: Record<string, MlSquareDebug> } | undefined;
  if (!raw?.squares) return null;
  return raw.squares;
}

File: apps/api/app/api/schemas/detect_board.py
"""Detect-board API schemas."""

import base64

from pydantic import BaseModel, Field

from app.api.schemas.common import PointSchema


class SquareInfoSchema(BaseModel):
    name: str = Field(..., description="Algebraic square name, e.g. e4")
    filename: str = Field(..., description="Crop filename, e.g. e4.png")
    cell_bbox: list[int] = Field(..., min_length=4, max_length=4)
    crop_bbox: list[int] = Field(..., min_length=4, max_length=4)
    url: str
    label: str = "empty"
    confidence: float = Field(0.0, ge=0.0, le=1.0)
    occupied: bool = False


class DetectionDebugSchema(BaseModel):
    original_base64: str | None = None
    detected_lines_base64: str | None = None
    intersections_base64: str | None = None
    mesh_base64: str | None = None
    rectified_board_base64: str | None = None
    rectified_upscaled_base64: str | None = None
    square_extraction_base64: str | None = None
    crop_quality_base64: str | None = None
    occupancy_base64: str | None = None
    occupancy_detail_base64: str | None = None
    classifier_confidence_base64: str | None = None
    ml_occupancy_base64: str | None = None
    ml_piece_top1_base64: str | None = None
    ml_onnx_occ_crops_base64: str | None = None
    ml_onnx_piece_crops_base64: str | None = None
    ml_detail_base64: str | None = None
    fen_candidates_base64: str | None = None
    mesh_quality_base64: str | None = None
    final_board_base64: str | None = None
    grid_debug_extreme_base64: str | None = None
    dataset_squares_base64: str | None = None


class BoardMatrixCellSchema(BaseModel):
    label: str
    confidence: float = Field(..., ge=0.0, le=1.0)


class DetectBoardResponse(BaseModel):
    job_id: str
    corners: list[PointSchema] = Field(..., min_length=4, max_length=4)
    homography: list[list[float]]
    confidence: float = Field(..., ge=0.0, le=1.0)
    original_width: int = Field(..., gt=0)
    original_height: int = Field(..., gt=0)
    output_width: int = Field(..., gt=0)
    output_height: int = Field(..., gt=0)
    warped_image_base64: str = Field(..., description="JPEG-encoded top-down board")
    squares: list[SquareInfoSchema] = Field(..., min_length=64, max_length=64)
    fen: str | None = Field(None, description="Best-effort FEN from classification (may be invalid)")
    interactive_fen: str | None = Field(None, description="Legal FEN for interactive board — null if not validated")
    fen_confidence: float | None = Field(None, ge=0.0, le=1.0)
    fen_valid: bool | None = None
    board_ready: bool = False
    board_matrix: list[list[BoardMatrixCellSchema]] | None = Field(
        None, description="Validated 8×8 piece matrix (rank 8 → rank 1)"
    )
    orientation: str | None = None
    debug: DetectionDebugSchema | None = None
    debug_overlay_base64: str | None = None
    debug_montage_base64: str | None = None
    processing_ms: int = Field(..., ge=0)
    metadata: dict

    @classmethod
    def from_detection(
        cls,
        *,
        job_id: str,
        corners: list[tuple[float, float]],
        homography: list[list[float]],
        confidence: float,
        original_width: int,
        original_height: int,
        output_width: int,
        output_height: int,
        warped_jpeg: bytes,
        squares: list[SquareInfoSchema],
        fen: str | None,
        interactive_fen: str | None,
        fen_confidence: float | None,
        fen_valid: bool | None,
        board_ready: bool,
        board_matrix: list[list[dict[str, str | float]]] | None,
        orientation: str | None,
        debug_overlay_jpeg: bytes | None,
        debug_montage_jpeg: bytes | None,
        debug_jpegs: dict[str, bytes] | None,
        processing_ms: int,
        metadata: dict,
    ) -> DetectBoardResponse:
        return cls(
            job_id=job_id,
            corners=[PointSchema(x=x, y=y) for x, y in corners],
            homography=homography,
            confidence=confidence,
            original_width=original_width,
            original_height=original_height,
            output_width=output_width,
            output_height=output_height,
            warped_image_base64=base64.b64encode(warped_jpeg).decode("ascii"),
            squares=squares,
            fen=fen,
            interactive_fen=interactive_fen,
            fen_confidence=fen_confidence,
            fen_valid=fen_valid,
            board_ready=board_ready,
            board_matrix=_map_board_matrix(board_matrix),
            orientation=orientation,
            debug=_build_debug_schema(debug_jpegs),
            debug_overlay_base64=_encode(debug_overlay_jpeg),
            debug_montage_base64=_encode(debug_montage_jpeg),
            processing_ms=processing_ms,
            metadata=metadata,
        )


def _map_board_matrix(
    raw: list[list[dict[str, str | float]]] | None,
) -> list[list[BoardMatrixCellSchema]] | None:
    if raw is None:
        return None
    return [
        [BoardMatrixCellSchema(label=str(cell["label"]), confidence=float(cell["confidence"])) for cell in row]
        for row in raw
    ]


def _build_debug_schema(debug_jpegs: dict[str, bytes] | None) -> DetectionDebugSchema | None:
    if not debug_jpegs:
        return None
    return DetectionDebugSchema(
        original_base64=_encode(debug_jpegs.get("original")),
        detected_lines_base64=_encode(debug_jpegs.get("detected_lines")),
        intersections_base64=_encode(debug_jpegs.get("intersections")),
        mesh_base64=_encode(debug_jpegs.get("mesh")),
        rectified_board_base64=_encode(debug_jpegs.get("rectified_board")),
        rectified_upscaled_base64=_encode(debug_jpegs.get("rectified_upscaled")),
        square_extraction_base64=_encode(debug_jpegs.get("square_extraction")),
        crop_quality_base64=_encode(debug_jpegs.get("crop_quality")),
        occupancy_base64=_encode(debug_jpegs.get("occupancy")),
        occupancy_detail_base64=_encode(debug_jpegs.get("occupancy_detail")),
        classifier_confidence_base64=_encode(debug_jpegs.get("classifier_confidence")),
        ml_occupancy_base64=_encode(debug_jpegs.get("ml_occupancy")),
        ml_piece_top1_base64=_encode(debug_jpegs.get("ml_piece_top1")),
        ml_onnx_occ_crops_base64=_encode(debug_jpegs.get("ml_onnx_occ_crops")),
        ml_onnx_piece_crops_base64=_encode(debug_jpegs.get("ml_onnx_piece_crops")),
        ml_detail_base64=_encode(debug_jpegs.get("ml_detail")),
        fen_candidates_base64=_encode(debug_jpegs.get("fen_candidates")),
        mesh_quality_base64=_encode(debug_jpegs.get("mesh_quality")),
        final_board_base64=_encode(debug_jpegs.get("final_board")),
        grid_debug_extreme_base64=_encode(debug_jpegs.get("grid_debug_extreme")),
        dataset_squares_base64=_encode(debug_jpegs.get("dataset_squares")),
    )


def _encode(data: bytes | None) -> str | None:
    if data is None:
        return None
    return base64.b64encode(data).decode("ascii")
